draft/0005-impersonation.md

```diff
@@ -0,0 +1,260 @@
+
+User impersonation is a feature that allows an authorized user to assume the identity of another user within the system. This capability can be useful for various reasons such as troubleshooting, providing support, and performing administrative tasks. This document aims to explore the design and implementation of user impersonation in Logto.
+
+## Motivation
+
+Logto is an enterprise ready identity provider, and it is essential to provide user impersonation capabilities to support various use cases. For example, a support engineer may need to troubleshoot an issue on behalf of a user, or an administrator may need to verify the permissions of a user. User impersonation can help to streamline these processes and improve the overall user experience.
+
+## Token exchange
+
+The user impersonation feature will be implemented using the OAuth 2.0 [token exchange flow](https://www.rfc-editor.org/rfc/rfc8693.html). The token exchange flow allows an authorized user to exchange their access token for a new access token that represents the identity of another user. The new access token can then be used to make API requests on behalf of the impersonated user.
+
+### Request
+
+The client makes a token exchange request to the tenant's token endpoint with a special grant type using the HTTP POST method. The following parameters are included in the HTTP request entity-body using the `application/x-www-form-urlencoded` format.
+
+1. `grant_type`: REQUIRED. The value of this parameter must be `urn:ietf:params:oauth:grant-type:token-exchange` indicates that a token exchange is being performed.
+2. `resource`: OPTIONAL. The resource indicator, the same as other token requests.
+3. `scope`: OPTIONAL. The requested scopes, the same as other token requests.
+4. `subject_token`: REQUIRED. A security token that represents the identity of the party on behalf of whom the request is being made. See the next section for more details.
+5. `subject_token_type`: REQUIRED. The type of the security token provided in the `subject_token` parameter. The value of this parameter must be `urn:ietf:params:oauth:token-type:access_token`.
+
+### Response
+
+If the token exchange request is successful, the tenant's token endpoint returns an access token that represents the identity of the impersonated user. The response includes the following parameters in the HTTP response entity-body using the `application/json` format.
+
+1. `access_token`: REQUIRED. The access token that represents the identity of the impersonated user.
+2. `issued_token_type`: REQUIRED. The type of the issued token. The value of this parameter must be `urn:ietf:params:oauth:token-type:access_token`.
+3. `token_type`: REQUIRED. The type of the token. The value of this parameter must be `Bearer`.
+4. `expires_in`: REQUIRED. The lifetime in seconds of the access token.
+5. `scope`: OPTIONAL. The scopes of the access token.
+6. `refresh_token`: OPTIONAL. The refresh token that can be used to obtain a new access token.
```

Comment:
is it needed to use `refresh_token` grant for upcoming token exchanges? if so, better to specify

Reply:
Updated